R/ard_categorical.R

```diff
@@ -154,117 +161,287 @@ ard_categorical <- function(data, variables, by = NULL, strata = NULL,
+  if (is_empty(variables)) return(list())
+  # if no by/strata and no denominator (or column), then use number of non-missing in variable
+  if ((is.null(denominator) || isTRUE(denominator %in% "column")) && is_empty(c(by, strata))) {
+    lst_denominator <-
+      lapply(
+        variables,
+        function(variable) dplyr::tibble(...ard_N... = sum(!is.na(data[[variable]])))
+      ) |>
+      stats::setNames(variables)
+  }
+  # if by/strata present and no denominator (or column), then use number of non-missing variable
+  else if (is.null(denominator) || isTRUE(denominator %in% "column")) {
+    lst_denominator <-
+      lapply(
+        variables,
+        function(variable) {
+          .table_as_df(
+            tidyr::drop_na(data, all_of(c(by, strata, variable))),
+            by = by,
+            strata = strata,
+            count_column = "...ard_N..."
+          )
+        }
+      ) |>
+      stats::setNames(variables)
+  }
+  # if user passed a data frame WITHOUT the counts pre-specified and no by/strata
+  else if (is.data.frame(denominator) &&
+           !"...ard_N..." %in% names(denominator) &&
+           is_empty(intersect(c(by, strata), names(denominator)))) {
+    lst_denominator <-
+      rep_named(
+        variables,
+        list(dplyr::tibble(...ard_N... = nrow(denominator)))
+      )
+  }
+  # if user passed a data frame WITHOUT the counts pre-specified
+  else if (is.data.frame(denominator) && !"...ard_N..." %in% names(denominator)) {
+    lst_denominator <-
+      rep_named(
+        variables,
+        list(
+          .table_as_df(
+            tidyr::drop_na(denominator, any_of(c(by, strata))),
+            by = intersect(by, names(denominator)),
+            strata = intersect(strata, names(denominator)),
+            count_column = "...ard_N..."
+          )
+        )
+      )
+  }
+  # if user requested cell percentages
+  else if (isTRUE(denominator %in% "cell")) {
+    lst_denominator <-
+      lapply(
+        variables,
+        function(variable) {
+          dplyr::tibble(
+            ...ard_N... =
+              tidyr::drop_na(data, all_of(c(by, strata, variable))) |> nrow()
+          )
+        }
+      ) |>
+      stats::setNames(variables)
+  }
+  # if user requested row percentages
+  else if (isTRUE(denominator %in% "row")) {
+    lst_denominator <-
+      lapply(
+        variables,
+        function(variable) {
+          .table_as_df(
+            tidyr::drop_na(data, all_of(c(by, strata, variable))),
+            variable = variable,
+            count_column = "...ard_N..."
+          )
+        }
+      ) |>
+      stats::setNames(variables)
+  }
+  # if user passed a single integer
+  else if (is_scalar_integerish(denominator)) {
+    lst_denominator <-
+      rep_named(
+        variables,
+        list(dplyr::tibble(...ard_N... = as.integer(denominator)))
+      )
+  }
+  # if user passed a data frame WITH the counts pre-specified
```

Comment:
whew!! great job covering all the conditions! 
just one comment here that this part will capture user-supplied df's that do not have all levels of by/strata represented (I think the non-represented ones will just result NA N&p in the end)...do we want to have a check for this?

Reply:
That's a good point! previously, we had a check that a passed data frame was required to contain all the by variables (not the strata), but I was thinking there are so many situations out there where the denom needs to be this or that, so i removed it. 

The way it's currently written, the N is the number of rows within each by/strata combo of the passed data frame, and if those variables do not appear, it's the number of rows in the data frame.

If you don't mind, let's leave it the way it is, and we can wait to hear how users feel about the implementation?

Reply:
oh wait, i misread your comment. like if they pass ADSL with the ARM column and only 1 level is present....YES, we should have a check for this!

Reply:
Thank you! Check has been added!

![image](https://github.com/insightsengineering/cards/assets/26774684/a62bcf15-5119-4845-b195-e069f45dc422)